keras_spiking/constraints.py

```diff
@@ -0,0 +1,36 @@
+"""
+Custom constraints for weight tensors in Keras models.
+"""
+
+import tensorflow as tf
+
+
+class Mean(tf.keras.constraints.Constraint):
+    """Constrains weight tensors to be their mean.
+
+    Parameters
+    ----------
+    axis : int
+        Axis used to compute the mean and repeat its value. Defaults to the last axis.
+    non_neg : boolean
+        Whether to constrain the resulting mean to be non-negative as well.
```

Comment:
We should probably just build this into the Lowpass layer (i.e., pass `tau` though a `relu` nonlinearity). Otherwise it'd be easy for people to accidentally shoot themselves in the foot by changing `tau_constraint`, and not realizing that they've lost the important non-negative property.

Reply:
I think this would be a breaking change potentially as gradient descent could behave differently if the clipping is happening inside of the inference call on an intermediate value rather than directly on the trained parameter after each gradient update. In the former case the underlying value of `tau_var` would be allowed to visit arbitrarily small negative values and potentially take a while to come back to positive, while the latter case it would always stay nonnegative and immediately be able to return to positive values if the gradient changes direction. So if this is the route being taken (clip tau in the inference call) I'd expect it to be a separate PR with a major version bump, and some checks to make sure it doesn't make it harder to train any important models/examples.

Reply:
Thinking about this more and we could always do both in the mean time (keep constraining it as we are, and then additionally clip it as an extra safety precaution)? Might appear a little confusing/redundant though.

Reply:
I suspect the difference between the two approaches would be minor. The underlying value of `tau_var` could drift slightly into the negative due to momentum terms in the optimizer, but I wouldn't expect that to be a major effect (that being said, I've never explicitly looked into it). We could also use something like `softplus` instead of `relu`, which would provide a more gradual thresholding. That would also be a breaking change as it'd change the effective value of `tau_var` in existing models, but we're in alpha so I'm not too worried about introducing breaking changes (as long as they're clearly marked by version changes and changelogs).

Reply:
I looked into this a bit and I think you are right that the difference is basically not perceptable. I was leaning on some experience I had with a similar situation when training `1 / theta` in the LMU (I was seeing the underlying value become negative, even with clipping) but I think that was different enough that the effect doesn't apply here. I'll add that change to this PR.